move javascript to the end of body tag.

wuboy0307 opened the pull request:
We should move javascript to the end of body tag for better performance

Please see discussion here.
http://stackoverflow.com/questions/1220956/move-jquery-to-the-end-of-body-tag

wuboy0307 committed:
move javascript to the end of body tag.
http://stackoverflow.com/questions/1220956/move-jquery-to-the-end-of-body-tag

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://wuboy0307-move-js.spree-spree-9ed724b77adec04c49c2.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://wuboy0307-move-js.spree-spree-9ed724b77adec04c49c2.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

JDutil commented on frontend/app/views/spree/shared/_foot.html.erb:
```diff
@@ -0,0 +1,23 @@
+-->
+
+<!-- load page java script after load all library-->
+<% if conten_for?(:foot) %>
```
content_for to fix specs

JDutil commented:
This has been attempted before and we had to revert it due to some issues with turbolinks not working properly.  Not sure if that has been fixed yet or not, but would be good to do this if we can.

wuboy0307 commented:
What's the detail of the turbolinks issue? I modified this and use them in my project and it works fine. Both in development and production environment.

By the way, this file in spree_i18n should also modified as well.
https://github.com/spree-contrib/spree_i18n/blob/master/app/overrides/spree/layouts/spree_application/add_js_params.html.erb.deface

jasonfb commented on frontend/app/views/spree/shared/_foot.html.erb:
```diff
@@ -0,0 +1,23 @@
+
+<!-- load page java script after load all library-->
+<% if content_for?(:foot) %>
+    <%= yield :foot %>
```
this is a good pattern, I use a pattern similar to this as well. However, it means that all inline javascript in Spree and your custom app (which is really a code smell anyway IMHO) needs to be moved into a content_for(:foot) block

I think various parts of spree may be littered still with inline javascript.
